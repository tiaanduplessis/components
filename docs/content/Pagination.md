---
title: Pagination
---
import State from '../components/State'

Use the pagination component to create a connected set of links that lead to related pages (for example, previous, next, or page numbers).

## Basic example

The pagination component only requires two properties to render: `pageCount`, which is the total number of pages, and `currentPage`, which is the currently selected page number (which should be managed by the consuming application).

```jsx live
<Pagination
  pageCount={15}
  currentPage={2}
  onPageChange={e => e.preventDefault()}
/>
```

However, to handle state changes when the user clicks a page, you also need to pass `onPageChange`, which is a function that takes a click event and page number as an argument:

```javascript
type PageChangeCallback = (evt: React.MouseEvent, page: number) => void
```

By default, clicking a link in the pagination component will cause the browser to navigate to the URL specified by the page. To cancel navigation and handle state management on your own, you should call `preventDefault` on the event, as in this example:

```jsx live
<State default={1}>
  {([page, setPage]) => {
    const totalPages = 15
    const onPageChange = (evt, page) => {
      evt.preventDefault()
      setPage(page)
    }

    return (
      <BorderBox p={2}>
        <Box>Current page: {page} / {totalPages}</Box>
        <Pagination
          pageCount={totalPages}
          currentPage={page}
          onPageChange={onPageChange}
        />
      </BorderBox>
    )
  }}
</State>
```

## Customizing link URLs

To customize the URL generated for each link, you can pass a function to the `hrefBuilder` property. The function should take a page number as an argument and return a URL to use for the link.

```javascript
type HrefBuilder = (page: number) => string
```

```jsx live
<State default={'(nothing clicked yet)'}>
  {([lastUrl, setLastUrl]) => {
    const onPageChange = (evt, page) => {
      evt.preventDefault()
      setLastUrl(evt.target.href)
    }
    const hrefBuilder = (page) => {
      return `https://example.com/pages/${page}`
    }

    return (
      <BorderBox p={2}>
        <Box>The last URL clicked was: {lastUrl}</Box>
        <Pagination
          pageCount={15}
          currentPage={2}
          onPageChange={onPageChange}
          hrefBuilder={hrefBuilder}
        />
      </BorderBox>
    )
  }}
</State>
```

## Customizing which pages are shown

Two props control how many links are displayed in the pagination container at any given time. `marginPageCount` controls how many pages are guaranteed to be displayed on the left and right of the component; `surroundingPageCount` controls how many pages will be displayed to the left and right of the current page.

```jsx live
<Pagination
  pageCount={20}
  currentPage={10}
  marginPageCount={1}
  surroundingPageCount={2}
  onPageChange={e => e.preventDefault()}
/>
```

The algorithm tries to minimize the amount the component shrinks and grows as the user changes pages; for this reason, if any of the pages in the margin (controlled via `marginPageCount`) intersect with pages in the center (controlled by `surroundingPageCount`), the center section will be shifted away from the margin. Consider the following examples, where pages one through six are shown when any of the first four pages are selected. Only when the fifth page is selected and there is a gap between the margin pages and the center pages does a break element appear.

```jsx live
<Box>
  {[1, 2, 3, 4, 5].map(page => (
    <Pagination
      pageCount={20}
      currentPage={page}
      marginPageCount={1}
      surroundingPageCount={2}
      onPageChange={e => e.preventDefault()}
    />
  ))}
</Box>
```

### Previous/next pagination

To hide all the page numbers and create a simple pagination container with just the Previous and Next buttons, set `showPages` to `false`.

```jsx live
<State default={1}>
  {([page, setPage]) => {
    const totalPages = 10
    const onPageChange = (evt, page) => {
      evt.preventDefault()
      setPage(page)
    }

    return (
      <BorderBox p={2}>
        <Box>Current page: {page} / {totalPages}</Box>
        <Pagination
          pageCount={totalPages}
          currentPage={page}
          onPageChange={onPageChange}
          showPages={false}
        />
      </BorderBox>
    )
  }}
</State>
```

## System props

Pagination components get `COMMON` system props. Read our [System Props](/system-props) doc page for a full list of available props.

## Component props

| Name | Type | Default | Description |
| :- | :- | :-: | :- |
| currentPage | Number | | **Required.** The currently selected page. |
| hrefBuilder | Function | `#${page}` | A function to generate links based on page number. |
| marginPageCount | Number | 1 | How many pages to always show at the left and right of the component. |
| onPageChange | Function | no-op | Called with event and page number when a page is clicked. |
| pageCount | Number | | **Required.** The total number of pages. |
| showPages | Boolean | `true` | Whether or not to show the individual page links. |
| surroundingPageCount | Number | 2 | How many pages to display on each side of the currently selected page. |

## Theming

The following snippet shows the properties in the theme that control the styling of the pagination component:

```javascript
{
  // ... rest of theme ...
  pagination: {
    borderRadius,
    spaceBetween,
    colors: {
      normal: {
        fg
      },
      disabled: {
        fg,
        border
      },
      hover: {
        border
      },
      selected: {
        fg,
        bg,
        border
      },
      active: {
        border
      },
      nextPrevious: {
        fg
      }
    }
  }
}
```
