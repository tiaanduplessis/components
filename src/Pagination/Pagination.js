import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'
import sx from '../sx'
import {get, COMMON} from '../constants'
import theme from '../theme'
import Box from '../Box'
import {buildPaginationModel, buildComponentData} from './model'

const Page = styled.a`
  display: inline-block;
  min-width: 32px;
  padding: 5px 10px;
  font-style: normal;
  line-height: 20px;
  color: ${get('pagination.colors.normal.fg')};
  text-align: center;
  white-space: nowrap;
  vertical-align: middle;
  cursor: pointer;
  user-select: none;
  text-decoration: none;

  margin-right: ${get('pagination.spaceBetween')};

  &:last-child {
    margin-right: 0;
  }

  border: ${get('borderWidths.1')} solid transparent;
  border-radius: ${get('pagination.borderRadius')};
  transition: border-color 0.2s cubic-bezier(0.3, 0, 0.5, 1);

  &:hover,
  &:focus {
    text-decoration: none;
    border-color: ${get('pagination.colors.hover.border')};
    outline: 0;
    transition-duration: 0.1s;
  }

  &:active {
    border-color: ${get('pagination.colors.active.border')};
  }

  &[rel='prev'],
  &[rel='next'] {
    color: ${get('pagination.colors.nextPrevious.fg')};
  }

  &[aria-current],
  &[aria-current]:hover {
    color: ${get('pagination.colors.selected.fg')};
    background-color: ${get('pagination.colors.selected.bg')};
    border-color: ${get('pagination.colors.selected.border')};
  }

  &[aria-disabled],
  &[aria-disabled]:hover {
    color: ${get('pagination.colors.disabled.fg')}; // check
    cursor: default;
    border-color: ${get('pagination.colors.disabled.border')};
  }

  @supports (clip-path: polygon(50% 0, 100% 50%, 50% 100%)) {
    &[rel='prev']::before,
    &[rel='next']::after {
      display: inline-block;
      width: 16px;
      height: 16px;
      vertical-align: text-bottom;
      content: '';
      background-color: currentColor;
    }

    // chevron-left
    &[rel='prev']::before {
      margin-right: ${get('pagination.spaceBetween')};
      clip-path: polygon(
        9.8px 12.8px,
        8.7px 12.8px,
        4.5px 8.5px,
        4.5px 7.5px,
        8.7px 3.2px,
        9.8px 4.3px,
        6.1px 8px,
        9.8px 11.7px,
        9.8px 12.8px
      );
    }

    // chevron-right
    &[rel='next']::after {
      margin-left: ${get('pagination.spaceBetween')};
      clip-path: polygon(
        6.2px 3.2px,
        7.3px 3.2px,
        11.5px 7.5px,
        11.5px 8.5px,
        7.3px 12.8px,
        6.2px 11.7px,
        9.9px 8px,
        6.2px 4.3px,
        6.2px 3.2px
      );
    }
  }

  ${COMMON};
`

function usePaginationPages({
  theme,
  pageCount,
  currentPage,
  onPageChange,
  hrefBuilder,
  marginPageCount,
  showPages,
  surroundingPageCount
}) {
  const pageChange = React.useCallback(n => e => onPageChange(e, n), [onPageChange])

  const model = React.useMemo(() => {
    return buildPaginationModel(pageCount, currentPage, showPages, marginPageCount, surroundingPageCount)
  }, [pageCount, currentPage, showPages, marginPageCount, surroundingPageCount])

  const children = React.useMemo(() => {
    return model.map(page => {
      const {props, key, content} = buildComponentData(page, hrefBuilder, pageChange(page.num))
      return (
        <Page {...props} key={key} theme={theme}>
          {content}
        </Page>
      )
    })
  }, [model, hrefBuilder, pageChange, theme])

  return children
}

const PaginationContainer = styled.nav`
  margin-top: 20px;
  margin-bottom: 15px;
  text-align: center;
  ${sx};
`

function Pagination({
  theme,
  pageCount,
  currentPage,
  onPageChange,
  hrefBuilder,
  marginPageCount,
  showPages,
  surroundingPageCount,
  ...rest
}) {
  const pageElements = usePaginationPages({
    theme,
    pageCount,
    currentPage,
    onPageChange,
    hrefBuilder,
    marginPageCount,
    showPages,
    surroundingPageCount
  })
  return (
    <PaginationContainer aria-label="Pagination" {...rest} theme={theme}>
      <Box display="inline-block" theme={theme}>
        {pageElements}
      </Box>
    </PaginationContainer>
  )
}

function defaultHrefBuilder(pageNum) {
  return `#${pageNum}`
}

function noop() {}

Pagination.propTypes = {
  currentPage: PropTypes.number.isRequired,
  hrefBuilder: PropTypes.func,
  marginPageCount: PropTypes.number,
  onPageChange: PropTypes.func,
  pageCount: PropTypes.number.isRequired,
  showPages: PropTypes.bool,
  surroundingPageCount: PropTypes.number,
  ...COMMON.propTypes,
  ...sx.propTypes
}

Pagination.defaultProps = {
  hrefBuilder: defaultHrefBuilder,
  marginPageCount: 1,
  onPageChange: noop,
  showPages: true,
  surroundingPageCount: 2,
  theme
}

export default Pagination
