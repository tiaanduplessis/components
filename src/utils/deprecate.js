/* eslint-disable no-console */
import {useRef, useCallback} from 'react'

const noop = () => {}
// eslint-disable-next-line import/no-mutable-exports
let deprecate = null

if (__DEV__) {
  deprecate = ({name, message, version}) => {
    Deprecations.deprecate({name, message, version})
  }
} else {
  deprecate = noop
}

export {deprecate}

// eslint-disable-next-line import/no-mutable-exports
let useDeprecation = null

if (__DEV__) {
  useDeprecation = ({name, message, version}) => {
    const ref = useRef(false)
    const logDeprecation = useCallback(() => {
      if (!ref.current) {
        ref.current = true
        deprecate({name, message, version})
      }
    }, [name, message, version])

    return logDeprecation
  }
} else {
  useDeprecation = () => {
    return noop
  }
}

export {useDeprecation}

export class Deprecations {
  static get() {
    if (!Deprecations.instance) {
      Deprecations.instance = new Deprecations()
    }

    return Deprecations.instance
  }

  constructor() {
    this.deprecations = []
  }

  static deprecate({name, message, version}) {
    const msg = `WARNING! ${name} is deprecated and will be removed in version ${version}. ${message}`
    console.warn(msg)

    this.get().deprecations.push({name, message, version})
  }

  static getDeprecations() {
    return this.get().deprecations
  }

  static clearDeprecations() {
    this.get().deprecations.length = 0
  }
}
