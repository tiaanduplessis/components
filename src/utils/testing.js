/* eslint-disable jsdoc/require-param, jsdoc/require-returns, jsdoc/require-description-complete-sentence, jsdoc/match-description */
import React from 'react'
import {promisify} from 'util'
import renderer from 'react-test-renderer'
import enzyme from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import {ThemeProvider} from 'styled-components'
import {default as defaultTheme} from '../theme'

const readFile = promisify(require('fs').readFile)

export const COMPONENT_DISPLAY_NAME_REGEX = /^[A-Z][A-Za-z]+(\.[A-Z][A-Za-z]+)*$/

enzyme.configure({adapter: new Adapter()})

export function mount(component) {
  return enzyme.mount(component)
}

/**
 * Render the component (a React.createElement() or JSX expression)
 * into its intermediate object representation with 'type',
 * 'props', and 'children' keys
 *
 * The returned object can be matched with expect().toEqual(), e.g.
 *
 * ```js
 * expect(render(<Foo />)).toEqual(render(<div foo='bar' />))
 * ```
 */
export function render(component) {
  return renderer.create(component).toJSON()
}

/**
 * Render the component (a React.createElement() or JSX expression)
 * using react-test-renderer and return the root node
 * ```
 */
export function renderRoot(component) {
  return renderer.create(component).root
}

/**
 * Get the HTML class names rendered by the component instance
 * as an array.
 *
 * ```js
 * expect(renderClasses(<div className='a b' />))
 *   .toEqual(['a', 'b'])
 * ```
 */
export function renderClasses(component) {
  const {
    props: {className}
  } = render(component)
  return className ? className.trim().split(' ') : []
}

/**
 * Returns true if a node renders with a single class.
 */
export function rendersClass(node, klass) {
  return renderClasses(node).includes(klass)
}

export function renderWithTheme(node, theme = defaultTheme) {
  return render(<ThemeProvider theme={theme}>{node}</ThemeProvider>)
}

export function px(value) {
  return typeof value === 'number' ? `${value}px` : value
}

export function percent(value) {
  return typeof value === 'number' ? `${value}%` : value
}

export function renderStyles(node) {
  const {
    props: {className}
  } = render(node)
  return getComputedStyles(className)
}

export function getComputedStyles(className) {
  const div = document.createElement('div')
  div.className = className

  const computed = {}
  for (const sheet of document.styleSheets) {
    for (const rule of sheet.cssRules) {
      if (rule.type === 1) {
        readRule(rule, computed)
      } else if (rule.type === 4) {
        readMedia(rule)
      } else {
        // console.warn('rule.type =', rule.type)
      }
    }
  }

  return computed

  function readRule(rule, dest) {
    if (matchesSafe(div, rule.selectorText)) {
      const {style} = rule
      for (let i = 0; i < style.length; i++) {
        const prop = style[i]
        dest[prop] = style.getPropertyValue(prop)
      }
    } else {
      // console.warn('no match:', rule.selectorText)
    }
  }

  function readMedia(mediaRule) {
    const key = `@media ${mediaRule.media[0]}`
    // const dest = computed[key] || (computed[key] = {})
    const dest = {}
    for (const rule of mediaRule.cssRules) {
      readRule(rule, dest)
    }

    // Don't add media rule to computed styles
    // if no styles were actually applied
    if (Object.keys(dest).length > 0) {
      computed[key] = dest
    }
  }

  function matchesSafe(node, selector) {
    if (!selector) {
      return false
    }
    try {
      return node.matches(selector)
    } catch (error) {
      return false
    }
  }
}

/**
 * This provides a layer of compatibility between the render() function from
 * react-test-renderer and Enzyme's mount()
 */
export function getProps(node) {
  return typeof node.props === 'function' ? node.props() : node.props
}

export function getClassName(node) {
  return getProps(node).className
}

export function getClasses(node) {
  const className = getClassName(node)
  return className ? className.trim().split(/ +/) : []
}

export function loadCSS(path) {
  return readFile(require.resolve(path), 'utf8').then(css => {
    const style = document.createElement('style')
    style.setAttribute('data-path', path)
    style.textContent = css
    document.head.appendChild(style)
    return style
  })
}

export function unloadCSS(path) {
  const style = document.querySelector(`style[data-path="${path}"]`)
  if (style) {
    style.remove()
    return true
  }
}

// If a component requires certain props or other conditions in order
// to render without errors, you can pass a `toRender` function that
// returns an element ready to be rendered.
export function behavesAsComponent(Component, systemPropArray, toRender = null, options) {
  if (typeof toRender === 'object' && !options) {
    options = toRender
    toRender = null
  }
  options = options || {}

  const getElement = () => (toRender ? toRender() : <Component />)

  it('implements system props', () => {
    for (const systemProps of systemPropArray) {
      expect(Component).toImplementSystemProps(systemProps)
    }
  })

  if (!options.skipSx) {
    it('implements the sx prop', () => {
      expect(Component).toImplementSxProp()
    })

    it('implements sx prop behavior', () => {
      expect(getElement()).toImplementSxBehavior()
    })
  }

  if (!options.skipAs) {
    it('respects the as prop', () => {
      const As = React.forwardRef((_props, ref) => <div className="as-component" ref={ref} />)
      const elem = React.cloneElement(getElement(), {as: As})
      expect(render(elem)).toEqual(render(<As />))
    })
  }

  it('sets a valid displayName', () => {
    expect(Component.displayName).toMatch(COMPONENT_DISPLAY_NAME_REGEX)
  })

  it('sets the default theme', () => {
    expect(getElement()).toSetDefaultTheme()
  })

  it('renders consistently', () => {
    expect(render(getElement())).toMatchSnapshot()
  })
}

export function checkExports(path, exports) {
  it('has declared exports', () => {
    const mod = require(`../${path}`)
    expect(mod).toSetExports(exports)
  })
}
