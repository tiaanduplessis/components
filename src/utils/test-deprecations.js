import semver from 'semver'
import {Deprecations} from '../utils/deprecate'

const ourVersion = require('../../package.json').version

beforeEach(() => {
  Deprecations.clearDeprecations()
})

afterEach(() => {
  const deprecations = Deprecations.getDeprecations()

  for (const dep of deprecations) {
    if (semver.gte(ourVersion, dep.version)) {
      throw new Error(`Found a deprecation that should be removed in ${dep.version}`)
    }
  }
})
