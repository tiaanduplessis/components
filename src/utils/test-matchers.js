import React from 'react'
import 'jest-styled-components'
import {styleSheetSerializer} from 'jest-styled-components/serializer'
import theme from '../theme'
import {getClasses, mount, getComputedStyles, render} from './testing'

expect.addSnapshotSerializer(styleSheetSerializer)

const stringify = d => JSON.stringify(d, null, '  ')

/**
 * These are props that styled-system aliases for backwards compatibility.
 * For some reason, they don't show up in our toImplementSystemProps() matcher,
 * so we skip over them.
 */
const ALIAS_PROP_TYPES = ['w', 'align', 'justify', 'wrap']

expect.extend({
  toMatchKeys(obj, values) {
    return {
      pass: Object.keys(values).every(key => this.equals(obj[key], values[key])),
      message: () => `Expected ${stringify(obj)} to have matching keys: ${stringify(values)}`
    }
  },

  toHaveClass(node, klass) {
    const classes = getClasses(node)
    const pass = classes.includes(klass)
    return {
      pass,
      message: () => `expected ${stringify(classes)} to include: ${stringify(klass)}`
    }
  },

  toHaveClasses(node, klasses, only = false) {
    const classes = getClasses(node)
    const pass = only ? this.equals(classes.sort(), klasses.sort()) : klasses.every(klass => classes.includes(klass))
    return {
      pass,
      message: () => `expected ${stringify(classes)} to include: ${stringify(klasses)}`
    }
  },

  toImplementSystemProps(Component, propNames) {
    const propKeys = new Set(Object.keys(Component.propTypes))
    const expectedPropKeys = Object.keys(propNames.propTypes)
    const missing = expectedPropKeys.filter(key => !propKeys.has(key)).filter(key => !ALIAS_PROP_TYPES.includes(key))
    return {
      pass: missing.length === 0,
      message: () => `Missing prop${missing.length === 1 ? '' : 's'}: ${stringify(missing)}`
    }
  },

  toImplementSxProp(Component) {
    return {
      pass: !!Component.propTypes.sx,
      message: () => 'Missing sx propTypes'
    }
  },

  toImplementSxBehavior(element) {
    const mediaKey = '@media (max-width:123px)'
    const sxPropValue = {
      [mediaKey]: {
        color: 'red.5'
      }
    }

    const elem = React.cloneElement(element, {sx: sxPropValue})
    const rendered = render(elem)

    function checkStylesDeep(rendered) {
      const className = rendered.props ? rendered.props.className : ''
      const styles = getComputedStyles(className)
      if (styles[mediaKey] && styles[mediaKey].color) {
        return true
      } else if (rendered.children) {
        return rendered.children.some(child => checkStylesDeep(child))
      } else {
        return false
      }
    }

    return {
      pass: checkStylesDeep(rendered),
      message: () => 'sx prop values did not change styles of component nor of any sub-components'
    }
  },

  toSetDefaultTheme(Component) {
    let comp
    if (Component.type) {
      comp = Component
    } else {
      comp = <Component />
    }
    const wrapper = mount(comp)
    const pass = this.equals(wrapper.prop('theme'), theme)
    return {
      pass,
      message: () => 'default theme is not set'
    }
  },

  toSetExports(mod, expectedExports) {
    if (!Object.keys(expectedExports).includes('default')) {
      return {
        pass: false,
        message: () => "You must specify the module's default export"
      }
    }

    const seen = new Set()
    for (const exp of Object.keys(expectedExports)) {
      seen.add(exp)
      if (mod[exp] !== expectedExports[exp]) {
        if (!mod[exp] && !expectedExports[exp]) {
          continue
        }

        return {
          pass: false,
          message: () => `Module exported a different value from key '${exp}' than expected`
        }
      }
    }

    for (const exp of Object.keys(mod)) {
      if (seen.has(exp)) {
        continue
      }

      if (mod[exp] !== expectedExports[exp]) {
        return {
          pass: false,
          message: () => `Module exported an unexpected value from key '${exp}'`
        }
      }
    }

    return {
      pass: true,
      message: () => ''
    }
  }
})
