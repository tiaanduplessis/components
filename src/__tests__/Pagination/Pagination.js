import React from 'react'
import Pagination from '../../Pagination'
import {behavesAsComponent} from '../../utils/testing'
import {COMMON} from '../../constants'
import {render as HTMLRender, cleanup} from '@testing-library/react'
import {axe, toHaveNoViolations} from 'jest-axe'
import 'babel-polyfill'
expect.extend(toHaveNoViolations)

const reqProps = {pageCount: 10, currentPage: 1}
const comp = <Pagination {...reqProps} />

describe('Pagination', () => {
  behavesAsComponent(Pagination, [COMMON], () => comp)

  it('should have no axe violations', async () => {
    const {container} = HTMLRender(comp)
    const results = await axe(container, {
      rules: {
        // The skip-link rule has to do with entire documents
        // and is not relevant to this component.
        // See https://dequeuniversity.com/rules/axe/3.3/skip-link?application=axeAPI
        'skip-link': {
          enabled: false
        }
      }
    })
    expect(results).toHaveNoViolations()
    cleanup()
  })
})
