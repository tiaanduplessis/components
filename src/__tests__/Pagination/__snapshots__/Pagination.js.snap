// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination renders consistently 1`] = `
.c1 {
  display: inline-block;
  display: inline-block;
}

.c2 {
  display: inline-block;
  min-width: 32px;
  padding: 5px 10px;
  font-style: normal;
  line-height: 20px;
  color: #24292e;
  text-align: center;
  white-space: nowrap;
  vertical-align: middle;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  margin-right: 4px;
  border: 1px solid transparent;
  border-radius: 6px;
  -webkit-transition: border-color 0.2s cubic-bezier(0.3,0,0.5,1);
  transition: border-color 0.2s cubic-bezier(0.3,0,0.5,1);
}

.c2:last-child {
  margin-right: 0;
}

.c2:hover,
.c2:focus {
  -webkit-text-decoration: none;
  text-decoration: none;
  border-color: #eaecef;
  outline: 0;
  -webkit-transition-duration: 0.1s;
  transition-duration: 0.1s;
}

.c2:active {
  border-color: #eaecef;
}

.c2[rel='prev'],
.c2[rel='next'] {
  color: #0366d6;
}

.c2[aria-current],
.c2[aria-current]:hover {
  color: #fff;
  background-color: #0366d6;
  border-color: transparent;
}

.c2[aria-disabled],
.c2[aria-disabled]:hover {
  color: #d1d5da;
  cursor: default;
  border-color: transparent;
}

.c0 {
  margin-top: 20px;
  margin-bottom: 15px;
  text-align: center;
}

@supports (-webkit-clip-path:polygon(50% 0,100% 50%,50% 100%)) or (clip-path:polygon(50% 0,100% 50%,50% 100%)) {
  .c2[rel='prev']::before,
  .c2[rel='next']::after {
    display: inline-block;
    width: 16px;
    height: 16px;
    vertical-align: text-bottom;
    content: '';
    background-color: currentColor;
  }

  .c2[rel='prev']::before {
    margin-right: 4px;
    -webkit-clip-path: polygon( 9.8px 12.8px,8.7px 12.8px,4.5px 8.5px,4.5px 7.5px,8.7px 3.2px,9.8px 4.3px,6.1px 8px,9.8px 11.7px,9.8px 12.8px );
    clip-path: polygon( 9.8px 12.8px,8.7px 12.8px,4.5px 8.5px,4.5px 7.5px,8.7px 3.2px,9.8px 4.3px,6.1px 8px,9.8px 11.7px,9.8px 12.8px );
  }

  .c2[rel='next']::after {
    margin-left: 4px;
    -webkit-clip-path: polygon( 6.2px 3.2px,7.3px 3.2px,11.5px 7.5px,11.5px 8.5px,7.3px 12.8px,6.2px 11.7px,9.9px 8px,6.2px 4.3px,6.2px 3.2px );
    clip-path: polygon( 6.2px 3.2px,7.3px 3.2px,11.5px 7.5px,11.5px 8.5px,7.3px 12.8px,6.2px 11.7px,9.9px 8px,6.2px 4.3px,6.2px 3.2px );
  }
}

<nav
  aria-label="Pagination"
  className="c0"
>
  <div
    className="c1"
    display="inline-block"
  >
    <span
      aria-disabled="true"
      className="c2"
    >
      Previous
    </span>
    <em
      aria-current="page"
      className="c2"
    >
      1
    </em>
    <a
      aria-label="Page 2"
      className="c2"
      href="#2"
      onClick={[Function]}
    >
      2
    </a>
    <a
      aria-label="Page 3"
      className="c2"
      href="#3"
      onClick={[Function]}
    >
      3
    </a>
    <a
      aria-label="Page 4"
      className="c2"
      href="#4"
      onClick={[Function]}
    >
      4
    </a>
    <a
      aria-label="Page 5"
      className="c2"
      href="#5"
      onClick={[Function]}
    >
      5
    </a>
    <a
      aria-label="Page 6"
      className="c2"
      href="#6"
      onClick={[Function]}
    >
      6
    </a>
    <span
      aria-disabled={true}
      className="c2"
    >
      â€¦
    </span>
    <a
      aria-label="Page 10"
      className="c2"
      href="#10"
      onClick={[Function]}
    >
      10
    </a>
    <a
      aria-label="Next Page"
      className="c2"
      href="#2"
      onClick={[Function]}
      rel="next"
    >
      Next
    </a>
  </div>
</nav>
`;
